<a href="https://colab.research.google.com/github/franckbizimana/Wamungu/blob/main/Pandas_Class_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Table of Content


1. **[Importing Modules (Pandas)](#pandas)**
<br><br>
2. **[Pandas DataFrame](#dataframes)**
<br><br>
3. **[Manipulating DataFrame](#dataframes)**
<br><br>
4. **[Reading Data from Different Sources](#reading_data)**



<a id="pandas"> </a>
# 1. Pandas

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> Pandas contain data structures and data manipulation tools designed for data cleaning and analysis.
<br><br>
                       Pandas is designed for working with tabular data.<br><br>

A module/library in python is simply a way to organize the code, and it contains either python classes or just functions.

### Quick look at functions

In [ ]:
# what are functions and arguments
def add(x,y):
    c = x+y
    return c

### First let's learn how modules work

In [ ]:
# Let's check where we have all our modules stored
import sys
print(sys.path)

['C:\\Users\\swapn\\Downloads', 'C:\\Users\\swapn\\anaconda3\\python311.zip', 'C:\\Users\\swapn\\anaconda3\\DLLs', 'C:\\Users\\swapn\\anaconda3\\Lib', 'C:\\Users\\swapn\\anaconda3', '', 'C:\\Users\\swapn\\anaconda3\\Lib\\site-packages', 'C:\\Users\\swapn\\anaconda3\\Lib\\site-packages\\win32', 'C:\\Users\\swapn\\anaconda3\\Lib\\site-packages\\win32\\lib', 'C:\\Users\\swapn\\anaconda3\\Lib\\site-packages\\Pythonwin']


The system will first look for the module in the `C:\\Users\\swapn\\Downloads` location. If it doesn't find it there it looks in `C:\\Users\\swapn\\anaconda3\\python311.zip` location, and so on.

In [ ]:
# create this math_operations.py file in C:\\Users\\swapn\\Downloads

def add_numbers(a, b):
    """Add two numbers."""
    return a + b

def subtract_numbers(a, b):
    """Subtract two numbers."""
    return a - b

def multiply_numbers(a=1, b=1):
    """Multiply two numbers with default arguments."""
    return a * b

In [ ]:
# import entire module
import math_operations
result = math_operations.add_numbers(3, 5)

In [ ]:
# import a specific function
from math_operations import subtract_numbers
result = subtract_numbers(10, 5)

In [ ]:
# import with alias
import math_operations as mo
result = mo.multiply_numbers(6, 7)

**How to install and import pandas?**<br>
1. Install pandas:<br><br>
`!pip install pandas`<br><br>
2. Import pandas:<br><br>
`import pandas as pd`

In [ ]:
#Check the list of base packages
!pip list

In [ ]:
# install pandas (one time only)
# !pip install pandas

In [ ]:
#import pandas 'library/package/modules'
import pandas as pd

 `as` is used as an alias in pandas. So from now on we will use `pd.` instead of `pandas.`

<br>
<span style="color:crimson">Always use libraries if they are freely available. It saves time, and those codes are already tested, debugged and optimized.</span>

<a id="dataframes"> </a>
# 2. Pandas DataFrames

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> A DataFrame is a tabular representation of data containing an ordered collection of columns, each of which can be a different type (numeric, string, boolean, and so on). <br><br>                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">To read data from a csv file</span>

In [ ]:
# read the example.csv file in a dataframe
df=pd.read_csv('missingdata.csv')

In [ ]:
# check the type
type(df)

pandas.core.frame.DataFrame

On checking the data type, we notice it is read as pandas data frame.

## <span style="color:darkgreen;">To print top & bottom rows of the data</span>

In [ ]:
#top 3 rows
df.head(3)

Age  Weight (in kg)  Height (in m)
0  45.0            60.0           1.35
1  12.0            43.0           1.21
2  54.0            78.0           1.50

By default, the `.head()` will display **first** five rows. However, we can set the desired number of rows to be displayed.

In [ ]:
#bottom 3 rows
df.tail(3)

Age  Weight (in kg)  Height (in m)
20  68.0            50.0           1.32
21   NaN            76.0           1.69
22  67.0            78.0           1.85

By default, the `.tail()` will display **last** five rows. However, we can set the desired number of rows to be displayed.

## <span style="color:darkgreen;">To obtain the dimensions of the data</span>

In [ ]:
# get the shape of the dataset
df.shape

(23, 3)

## <span style="color:darkgreen;">To know the data types of a data frame</span>

In [ ]:
# get all the data types
df.dtypes

Age               float64
Weight (in kg)    float64
Height (in m)     float64
dtype: object

We see the data type of each variable.

## <span style="color:darkgreen;">Print more information about the data</span>

In [ ]:
# All information together
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             22 non-null     float64
 1   Weight (in kg)  21 non-null     float64
 2   Height (in m)   22 non-null     float64
dtypes: float64(3)
memory usage: 680.0 bytes


We see this output gives the number of rows present in the data `RangeIndex: 23 entries, 0 to 22` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 3 columns)`.

Consider `Age 23 non-null int64` indicates that the column named 'Age' has 23 non-null observations having the data type 'int64'

And finally the memory used to save this dataframe is 680 bytes.

In [ ]:
# describe your data
df.describe()

Age  Weight (in kg)  Height (in m)
count  22.000000       21.000000      22.000000
mean   44.000000       59.285714       1.513636
std    20.914793       19.997857       0.221307
min    10.000000       21.000000       1.210000
25%    26.000000       45.000000       1.320000
50%    49.500000       65.000000       1.520000
75%    64.500000       76.000000       1.690000
max    75.000000       89.000000       1.850000

# `.loc` and `.iloc` methods

## <span style="color:darkgreen;">Indexing a dataframe using `.loc`</span>

`DataFrame.loc[]` is label-based method, which means that you have to specify rows and columns based on their row and column labels.

In [ ]:
#syntax
dataframe.loc[row-start:row-end, column-start:column-end]

In [ ]:
# @title Default title text


In [ ]:
# get first row with all columns
df.loc[0:22,"Age"]

0     45.0
1     12.0
2     54.0
3     26.0
4     68.0
5     21.0
6     10.0
7     57.0
8     75.0
9     32.0
10    23.0
11    34.0
12    55.0
13    23.0
14    56.0
15    67.0
16    26.0
17    56.0
18    67.0
19    26.0
20    68.0
21     NaN
22    67.0
Name: Age, dtype: float64

In [32]:
df.loc[0:22,["Age","Weight (in kg)"]]

Age  Weight (in kg)
0   45.0            60.0
1   12.0            43.0
2   54.0            78.0
3   26.0            65.0
4   68.0            50.0
5   21.0             NaN
6   10.0            32.0
7   57.0            34.0
8   75.0            23.0
9   32.0            21.0
10  23.0             NaN
11  34.0            65.0
12  55.0            89.0
13  23.0            45.0
14  56.0            76.0
15  67.0            78.0
16  26.0            65.0
17  56.0            74.0
18  67.0            78.0
19  26.0            65.0
20  68.0            50.0
21   NaN            76.0
22  67.0            78.0

In [33]:
df.loc[0:22,"Weight (in kg)"]

0     60.0
1     43.0
2     78.0
3     65.0
4     50.0
5      NaN
6     32.0
7     34.0
8     23.0
9     21.0
10     NaN
11    65.0
12    89.0
13    45.0
14    76.0
15    78.0
16    65.0
17    74.0
18    78.0
19    65.0
20    50.0
21    76.0
22    78.0
Name: Weight (in kg), dtype: float64

In [34]:
# get the value of Age in the 1st row
df.loc[0,"Age"]

45.0

## <span style="color:darkgreen;">Selecting multiple rows</span>

In [35]:
# select row 4, 7, 10
df.loc[[4,7,10]]

Age  Weight (in kg)  Height (in m)
4   68.0            50.0           1.32
7   57.0            34.0           1.61
10  23.0             NaN           1.50

In [36]:
df.loc[[4,7,10],["Age","Weight (in kg)"]]

Age  Weight (in kg)
4   68.0            50.0
7   57.0            34.0
10  23.0             NaN

In [37]:
df.loc[[2,12,19],["Age"]]

Age
2   54.0
12  55.0
19  26.0

We use two square brackets since we are passing a list of row numbers to be accessed.

## <span style="color:darkgreen;">Selecting a range of rows</span>

In [38]:
# select rows from 11 to 18
df.loc[11:18]

Age  Weight (in kg)  Height (in m)
11  34.0            65.0           1.76
12  55.0            89.0           1.65
13  23.0            45.0           1.75
14  56.0            76.0           1.69
15  67.0            78.0           1.85
16  26.0            65.0           1.21
17  56.0            74.0           1.69
18  67.0            78.0            NaN

## <span style="color:darkgreen;">Selecting the first column</span>

In [39]:
# select 1st column with all rows
df.loc[:,["Age"]]

Age
0   45.0
1   12.0
2   54.0
3   26.0
4   68.0
5   21.0
6   10.0
7   57.0
8   75.0
9   32.0
10  23.0
11  34.0
12  55.0
13  23.0
14  56.0
15  67.0
16  26.0
17  56.0
18  67.0
19  26.0
20  68.0
21   NaN
22  67.0

In [45]:
df.loc[:,["Weight (in kg)"]]

Weight (in kg)
0             60.0
1             43.0
2             78.0
3             65.0
4             50.0
5              NaN
6             32.0
7             34.0
8             23.0
9             21.0
10             NaN
11            65.0
12            89.0
13            45.0
14            76.0
15            78.0
16            65.0
17            74.0
18            78.0
19            65.0
20            50.0
21            76.0
22            78.0

To select the last column we use -1, to select the second last column we use -2

## <span style="color:darkgreen;">Select the first two columns</span>

In [46]:
# select the first two columns with all rows
df.loc[:,["Age","Weight (in kg)"]]

Age  Weight (in kg)
0   45.0            60.0
1   12.0            43.0
2   54.0            78.0
3   26.0            65.0
4   68.0            50.0
5   21.0             NaN
6   10.0            32.0
7   57.0            34.0
8   75.0            23.0
9   32.0            21.0
10  23.0             NaN
11  34.0            65.0
12  55.0            89.0
13  23.0            45.0
14  56.0            76.0
15  67.0            78.0
16  26.0            65.0
17  56.0            74.0
18  67.0            78.0
19  26.0            65.0
20  68.0            50.0
21   NaN            76.0
22  67.0            78.0

In [48]:
df.loc[:,["Height (in m)","Weight (in kg)"]]

Height (in m)  Weight (in kg)
0            1.35            60.0
1            1.21            43.0
2            1.50            78.0
3            1.21            65.0
4            1.32            50.0
5            1.52             NaN
6            1.65            32.0
7            1.61            34.0
8            1.24            23.0
9            1.52            21.0
10           1.50             NaN
11           1.76            65.0
12           1.65            89.0
13           1.75            45.0
14           1.69            76.0
15           1.85            78.0
16           1.21            65.0
17           1.69            74.0
18            NaN            78.0
19           1.21            65.0
20           1.32            50.0
21           1.69            76.0
22           1.85            78.0

## <span style="color:darkgreen;">Indexing a dataframe using `.iloc`</span>

`DataFrame.iloc[]` is integer position-based, so you have to specify rows and columns by their integer position values (0-based integer position).

**Note:** the row names are numbers

In [49]:
# using iloc select 'Age' & 'Weight (in kg)'
df.iloc[:,[0,2]]

Age  Height (in m)
0   45.0           1.35
1   12.0           1.21
2   54.0           1.50
3   26.0           1.21
4   68.0           1.32
5   21.0           1.52
6   10.0           1.65
7   57.0           1.61
8   75.0           1.24
9   32.0           1.52
10  23.0           1.50
11  34.0           1.76
12  55.0           1.65
13  23.0           1.75
14  56.0           1.69
15  67.0           1.85
16  26.0           1.21
17  56.0           1.69
18  67.0            NaN
19  26.0           1.21
20  68.0           1.32
21   NaN           1.69
22  67.0           1.85

In [51]:
df.iloc[0:10,[1,2]]

Weight (in kg)  Height (in m)
0            60.0           1.35
1            43.0           1.21
2            78.0           1.50
3            65.0           1.21
4            50.0           1.32
5             NaN           1.52
6            32.0           1.65
7            34.0           1.61
8            23.0           1.24
9            21.0           1.52

<a id="manipulatingDF"> </a>
# 3. Manipulating a Dataframe

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> CAUTION:<br>
                        1. DataFrame[column] works for any column name, but DataFrame.column only works when the column name is a valid Python variable name.<br>
                        2. New columns cannot be created with the ` data.BMI ` syntax.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">Adding a new column to the dataframe</span>

columns/variables/features mean the same

In [52]:
# create a new column BMI which is given by weight / H**2
df['BMI'] = df['Weight (in kg)'] / df['Height (in m)']**2

In [54]:
# check the shape of the data
df.shape

(23, 4)

## <span style="color:darkgreen;">Adding a new row to the dataframe</span>

In [55]:
# add a new row to the data with values [56, 76, 1.69, 26.609713] using .loc method
df.loc[23] = [56, 76, 1.69, 26.609713]

We see that a new row number 23 has be added to the data.

## <span style="color:darkgreen;">Sorting the dataframe</span>

In [56]:
pd??

In [ ]:
# sort the data frame on basis of 'Age' values, by default the values will get sorted in ascending order
# Note: 'ascending = False' will sort the data frame in descending order.


## <span style="color:darkgreen;">Droping Rows and Columns</span>

In [57]:
# create a copy of the dataframe
df_copy = df.copy()

In [58]:
# drop the BMI column
df_copy.drop(columns=['BMI'])

Age  Weight (in kg)  Height (in m)
0   45.0            60.0           1.35
1   12.0            43.0           1.21
2   54.0            78.0           1.50
3   26.0            65.0           1.21
4   68.0            50.0           1.32
5   21.0             NaN           1.52
6   10.0            32.0           1.65
7   57.0            34.0           1.61
8   75.0            23.0           1.24
9   32.0            21.0           1.52
10  23.0             NaN           1.50
11  34.0            65.0           1.76
12  55.0            89.0           1.65
13  23.0            45.0           1.75
14  56.0            76.0           1.69
15  67.0            78.0           1.85
16  26.0            65.0           1.21
17  56.0            74.0           1.69
18  67.0            78.0            NaN
19  26.0            65.0           1.21
20  68.0            50.0           1.32
21   NaN            76.0           1.69
22  67.0            78.0           1.85
23  56.0            76.0           1.69

In [59]:
# dropping the 23rd row
df_copy.drop(index=23)

Age  Weight (in kg)  Height (in m)        BMI
0   45.0            60.0           1.35  32.921811
1   12.0            43.0           1.21  29.369579
2   54.0            78.0           1.50  34.666667
3   26.0            65.0           1.21  44.395875
4   68.0            50.0           1.32  28.696051
5   21.0             NaN           1.52        NaN
6   10.0            32.0           1.65  11.753903
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
9   32.0            21.0           1.52   9.089335
10  23.0             NaN           1.50        NaN
11  34.0            65.0           1.76  20.983988
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
16  26.0            65.0           1.21  44.395875
17  56.0            74.0           1.69  25.909457
18  67.0            78.0            NaN        NaN
19  26.0            65.0           1.21  44.395875
20  68.0            50.0           1.32  28.696051
21   NaN            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358

## <span style="color:darkgreen;">Droping duplicates</span>

In [60]:
# Check if data has duplicates
df.duplicated().sum()

4

In [61]:
# drop duplicates from your data
df.drop_duplicates()

Age  Weight (in kg)  Height (in m)        BMI
0   45.0            60.0           1.35  32.921811
1   12.0            43.0           1.21  29.369579
2   54.0            78.0           1.50  34.666667
3   26.0            65.0           1.21  44.395875
4   68.0            50.0           1.32  28.696051
5   21.0             NaN           1.52        NaN
6   10.0            32.0           1.65  11.753903
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
9   32.0            21.0           1.52   9.089335
10  23.0             NaN           1.50        NaN
11  34.0            65.0           1.76  20.983988
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
17  56.0            74.0           1.69  25.909457
18  67.0            78.0            NaN        NaN
21   NaN            76.0           1.69  26.609713
23  56.0            76.0           1.69  26.609713

## <span style="color:darkgreen;">Checking for missing values</span>

Let's import a new dataset.

In [62]:
# Import missingdata.csv
mdata = pd.read_csv('missingdata.csv')
mdata.head(2)

Age  Weight (in kg)  Height (in m)
0  45.0            60.0           1.35
1  12.0            43.0           1.21

In [63]:
# check for nulls
mdata.info()
#mdata.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             22 non-null     float64
 1   Weight (in kg)  21 non-null     float64
 2   Height (in m)   22 non-null     float64
dtypes: float64(3)
memory usage: 680.0 bytes


The function `.isnull` check whether the data is missing. The `sum()` sums the number of 'True' values in the column. The final output gives the number of missing values in each column.

Here, we see there are 2 missing values in the 'weight' column and one missing value in other columns.

## <span style="color:crimson;">Take home exercise</span>

<a id="reading_data"> </a>
### Reading Data from Different Sources

Note that the files names are used as examples only. You can try importing your own files to execute the below examples.

**1. Read a `.xlsx` file**

`pd.read_excel('example.xlsx')`

**2. Read a `.txt` file**

`data = pd.read_csv('example.txt', sep="\t")`

**3. Read a `.zip` file**

`import zipfile
with zipfile.ZipFile('data.zip') as z:
    with z.open('example.csv') as f:
        file = pd.read_csv(f)
        print(file.head())`

**4. Read a `.html` file**

`df = pd.read_html('example.html', header=1, index_col=0)`

**5. Read a `.json` file**

`pd.read_json('example.json')`